format/composite.go

```diff
@@ -345,6 +263,84 @@ func (c *CompositeFormatter) Close(ctx context.Context) error {
 	return nil
 }
 
+// Hash takes anything that might affect the paths to be traversed, or how they are traversed, and adds it to a sha256
+// hash.
+// This can be used to determine if there has been a material change in config or setup that requires the cache
+// to be invalidated.
+func (c *CompositeFormatter) Hash() (string, error) {
+	h := sha256.New()
+
+	// start with the global excludes
+	h.Write([]byte(strings.Join(c.cfg.Excludes, " ")))
+	h.Write([]byte(strings.Join(c.cfg.Global.Excludes, " ")))
+
+	// sort formatters determinstically
+	formatters := make([]*Formatter, 0, len(c.formatters))
+	for _, f := range c.formatters {
+		formatters = append(formatters, f)
+	}
+
+	slices.SortFunc(formatters, formatterSortFunc)
+
+	// apply them to the hash
+	for _, f := range formatters {
+		if err := f.Hash(h); err != nil {
+			return "", fmt.Errorf("failed to hash formatter: %w", err)
+		}
+	}
+
+	// finalize
+	return hex.EncodeToString(h.Sum(nil)), nil
+}
+
+// BustCache compares the current Hash() of this formatter with the last entry stored in the db.
+// If it has changed, we remove all the path entries, forcing a fresh cache state.
+func (c *CompositeFormatter) BustCache(db *bolt.DB) error {
```

Comment:
I'm surprised to still see a function like this. Can't we just not do this work, and the right thing will happen?

Is this just a mechanism for keeping the cache small? Or is this a correctness thing?

If this is just a mechanism for keeping the cache small, I'd rather drop it in favor of implementing #456.

Reply:
It's a correctness thing. Consider a formatter's config: 

- if `Command` changes, the outcome of applying the formatter might differ. 
- Similarly, if the underlying executable that `Command` refers to has changed, the outcome of applying the formatter might differ. 
- if `Options` change, the outcome of applying the formatter might differ.
- if `Priority` changes, the outcome of applying the formatter in a particular sequence might be different. 
- If `Includes/Excludes` changes, there may be some new files to which we should apply the formatter.

Similarly, if the global excludes change, there may be new files to which we should apply the formatters.

We shouldn't wait until the underlying file changes before doing so; that's why we bust the cache. We could try to be more nuanced and expire only files that are affected. But the complexity of doing that versus the minimal cost of busting the cache is not worth it, in my opinion.

Reply:
It would be nice if it only busted the cache for the current formatter group instead of the whole cache.

Reply:
This is addressed in the latest changes I just pushed :slightly_smiling_face: